chain/network/src/peer/peer_actor.rs

```diff
@@ -247,7 +236,61 @@ impl PeerActor {
                 connection: None,
                 _connection_guard: connection_guard,
-            }
+            };
+            let addr = ctx.address();
+            let stats = pa.stats.clone();
+            // Event loop receiving and processing messages.
+            // After dispatching a message, loop starts parsing the next message immediately
+            // (i.e. it doesn't wait for the message to be processed).
+            // It uses a fixed small buffer allocated by BufReader.
+            // For each message it allocates a Vec with exact size of the message.
+            // TODO(gprusak): once borsh support is dropped, we can parse a proto
+            // directly from the stream.
+            ctx.spawn(
+                async move {
+                    const READ_BUFFER_CAPACITY: usize = 8 * 1024;
+                    let mut read = tokio::io::BufReader::with_capacity(READ_BUFFER_CAPACITY, read);
+                    loop {
+                        let n =
+                            read.read_u32_le().await.map_err(RecvLoopError::Receiving)? as usize;
+                        if n > NETWORK_MESSAGE_MAX_SIZE_BYTES {
+                            addr.do_send(RecvMessage(Err(ReasonForBan::Abusive)));
+                            return Err(RecvLoopError::MessageTooLarge {
+                                got_bytes: n,
+                                want_max_bytes: NETWORK_MESSAGE_MAX_SIZE_BYTES,
+                            });
+                        }
+                        let mut buf = Vec::new();
+                        buf.resize(n, 0);
+                        read.read_exact(&mut buf[..]).await.map_err(RecvLoopError::Receiving)?;
+                        stats.received_messages.fetch_add(1, Ordering::Relaxed);
+                        stats.received_bytes.fetch_add(n as u64, Ordering::Relaxed);
+                        addr.do_send(RecvMessage(Ok(buf)));
+                    }
+                }
+                .into_actor(&pa)
+                .then(move |res: Result<(), RecvLoopError>, act, ctx| {
+                    if let Err(err) = res {
+                        match err {
+                            RecvLoopError::Receiving(err)
+                                if err.kind() == io::ErrorKind::UnexpectedEof => {}
+                            err => error!(
+                                target: "network",
+                                ?err,
+                                "{}",
+                                act.peer_info,
+                            ),
+                        }
+                    }
+                    // Here we rely on the fact that this ctx.stop() is executed AFTER
+                    // all dispatched message are processed (i.e. we rely on in-order
+                    // execution of the actix queue), or at least first phase of processing
+                    // is done (since the RecvMessage handler can schedule next messages).
+                    ctx.stop();
```

Comment:
Alternatively, right now we have two exit points: here, and the ban in "on message". Maybe funnel everything through on message, so that it deals with three cases: valid message, io error, ban?

Reply:
done